Fallback Font Size Adjustment

qwerasd205 opened the pull request:
This PR changes `font.Collection` to automagically adjust the sizes of added fonts so that their metrics (specifically their ex height, or their ideograph character width if they have one) match the primary font. This is like [`font-size-adjust`](https://developer.mozilla.org/en-US/docs/Web/CSS/font-size-adjust) from CSS.

This is a big win for users who use mixed writing systems and rely heavily on fallback fonts. For example, in #7774 it's pointed out that CJK characters are not very well harmonized with existing Latin glyphs, well:
|Before (`main`)|After (this PR)|
|-|-|
|<img width="326" alt="image" src="https://github.com/user-attachments/assets/c11d372d-ec69-426d-b008-1f56a7430f23" />|<img width="326" alt="image" src="https://github.com/user-attachments/assets/efcb56ea-0572-481a-b632-a0b5cd170fa9" />|

This also improves our handling of the horizontal alignment of fallback glyphs. It's not an ideal solution; it only works for glyphs narrower than the cell width because it messes with ligatures if we include glyphs wider than the cell width; and most things would look better if the center were proportionally remapped based on the ratio from the glyph advance to the cell width, but that messes with glyphs designed to align vertically so it can't be done, instead the original advance width is centered in the cell width.

> [!NOTE]
> I can add a couple tests for the size adjust behavior (one for ex height, one for ic width) if you'd like.

### Future work
When working on this I was very bothered by the fact that our fallback fonts are not selected deterministically. That is, if a fallback with glyphs A and B is selected as the fallback for B because no other font has it, then it will automatically be used as the fallback for A as well even if other fonts would otherwise rank higher. This is a big problem if you have something like UniFont installed, since the moment that's loaded no other font will be loaded afterwards and you get very suboptimal selections for your glyphs.

So I'll have a think on how to go about making our fallbacks deterministic without killing performance (it's not a trivial problem unfortunately) and that will probably be my next area of work.

qwerasd205 committed:
font: remove unused fields from Glyph

We can reintroduce `advance` if we ever want to do proportional string
drawing, but we don't use it anywhere right now. And we also don't need
`sprite` anymore since that was just there to disable constraints for
sprites back when we did them on the GPU.

qwerasd205 committed:
font: more generic bearing adjustments

This generally adjusts the bearings of any glyph whose original advance
was narrower than the cell, which helps a lot with proportional fallback
glyphs so they aren't just left-aligned. This only applies to situations
where the glyph was originally narrower than the cell, so that we don't
mess up ligatures, and this centers the old advance width in the new one
rather than adjusting proportionally, because otherwise we can mess up
glyphs that are meant to align with others when placed vertically.

qwerasd205 committed:
font: adjust fallback font sizes to match primary metrics

This better harmonizes fallback fonts with the primary font by matching
the heights of lowercase letters. This should be a big improvement for
users who use mixed scripts and so rely heavily on fallback fonts.

mitchellh reviewed (approved):
I'll let you merge this one when you're ready in case there are any adjustments you wanted. Looks great!

qwerasd205 commented:
Thanks, I went to add a test and the test found a bug haha, got that fixed, cleaning it up and will update the PR then merge.

qwerasd205 committed:
fix(font): include line gap in `lineHeight` helper

qwerasd205 committed:
font: add test for size adjustment, fix small bug in resize

Previously produced very wrong values when calling Collection.setSize,
since it was assuming that the provided face had the same point size as
the primary face, which isn't true during resize-- so instead we just
have faces keep track of their set size, this is generally useful.

ken-wizhodl commented:
this PR literaly ruined Chinese charaters, please revert
